### PR TITLE
Immer 10

### LINKED ISSUE #867: Immer takes a long time to update the data.
## 🐛 Bug Report

Immer takes a long time to update the data.

## Link to repro

image

## To Reproduce

Chrome console:

```js
console.time('create');
const dataSource = Object.fromEntries(
  [...Array(4000).keys()].map((key) => [key, { key, data: { value: key } }]),
);
console.timeEnd('create');

console.time('update');
const result = produce(dataSource, (draft) => {
  draft[1000].data.value = 100;
});

console.timeEnd('update');
```

## Observed behavior

I checked the source code, the problem is in `shallowCopy`。If you use `Object.assign` instead, there is no performance problem

Time-consuming code:
```js
...
const descriptors = getOwnPropertyDescriptors(base)
...
return Object.create(Object.getPrototypeOf(base), descriptors)
```

## Expected behavior

Time shouldn't be so long

## Environment

We only accept bug reports against the latest Immer version.

- **Immer version:**
- [x] I filed this report against the _latest_ version of Immer
- [x] Occurs with `setUseProxies(true)`
- [ ] Occurs with `setUseProxies(false)` (ES5 only)

### PR DESCRIPTION
**Pull request by mweststrate**
# Release notes

* [breaking change] Immer 10 only supports modern browsers, that have support for `Proxy`, `Reflect`, `Symbol` and `Map` and `Set`.
* [breaking change] There is no longer a UMD build exposed (thanks [Mark Erikson](https://github.com/markerikson) for modernizing the build setup in #1032!
* [breaking change] getters and setters are ignored by default on plain object, as this is a very uncommon case and provides a significant performance boost (ca 33%, but depends a lot on the scenario). Fixes #867, #1012. Thanks [hrsh7th](https://github.com/hrsh7th) for implementing it in #941!
* [breaking change] Promise based reducers are no longer supported. Conceptually it is an anti pattern to hold on to drafts over time. If needed the old behavior can still be achieved by leveraging `createDraft` and `finishDraft`.
* [breaking change] ES5 mode (for legacy browsers) has been dropped. If your project relies on `enableES5()`, you SHOULD NOT upgrade Immer. `enableES5` has been removed.
* [breaking change] `produce` is no longer exposed as the `default` export. This improves eco system compatibility, and makes sure that there is only one correct way of doing things
* [breaking change] `enableAllPlugins` has been removed, use `enablePatches(); enableMapSet()` instead
* [breaking change] shortening the length of a JSON array now results in delete patches, rather than a mutation of the `length` property, in accordance with JSON spec. Thanks [kshramt](https://github.com/kshramt) for implementing this in #964!
* Immer is now an ESM package that can be directly imported into the browser. CJS should still work, UMD support has been removed.

Overall, there is a rough performance increase of 33% for Immer (and in some cases significantly higher), and a small overall package size reduction (TODO -> TODO bytes), with a small increase if only `produce` is imported: TODO -> TODO, or TODO -> TODO if everything is used. 

For more details, see #1015

## Migration steps

1. If you have any `enableES5()` call, don't migrate
1. When using getters/ setters icmw plain objects, call `useStrictShallowCopy(true)` at startup
1. Replace all default imports: Replace `import produce from "immer"` with `import {produce} from "immer"`
3. Replace all calls to `enableAllPlugins()` with `enablePatches(); enableMapSet();` to be more specific and smoothen future migrations. 
4. If any producer returned a Promise, refactor it to leverage `createDraft` instead. Roughly:

```javascript
const newState = await produce(oldState, recipe)

// becomes
const draft  = createDraft(oldState)
await recipe(draft)
const newState = finishDraft(draft)
```

### CODE REVIEW

File: __tests__/base.js [hunk 1]
```diff
@@ -890,10 +896,16 @@ function runBaseTest(name, autoFreeze, useListener) {
...
-				foo: true
-			})
+			if (useStrictShallowCopy) {
+				expect(nextState).toEqual({
+					[test]: true,
+					foo: true
+				})
+			} else {
+				expect(nextState).toEqual({
+					foo: true
```
Comment (mweststrate) [inline __tests__/base.js:906]: TODO: check why `[test]` is lost

File: src/core/current.ts [hunk 2]
```diff
@@ -28,10 +28,14 @@ function currentImpl(value: any): any {
...
-		copy = copyHelper(value, archType)
+		copy = copyHelper(
+			value,
+			archType,
+			state.scope_.immer_.useStrictShallowCopy_
+		)
 		state.finalized_ = false
 	} else {
-		copy = copyHelper(value, archType)
+		copy = copyHelper(value, archType, true)
```
Comment (mweststrate) [inline src/core/current.ts:38]: TODO: why `true`?

File: src/core/proxy.ts [hunk 3]
```diff
@@ -17,6 +17,7 @@ import {
 	createProxy,
 	ArchType
 } from "../internal"
+import {ImmerScope} from "./scope"
```
Comment (mweststrate) [inline src/core/proxy.ts:20]: TODO: import from internal

File: src/types/index.js.flow [hunk 4]
```diff
@@ -83,6 +83,13 @@ declare export function setAutoFreeze(autoFreeze: boolean): void
  */
 declare export function setUseProxies(useProxies: boolean): void
```
Comment (mweststrate) [inline src/types/index.js.flow:84]: TODO: remove

File: website/docs/api.md [hunk 5]
```diff
@@ -30,6 +30,7 @@ title: API overview
 | `produceWithPatches` | Works the same as `produce`, but instead of just returning the produced object, it returns a tuple, consisting of `[result, patches, inversePatches]`. | [Patches](./patches.mdx) |
 | `setAutoFreeze` | Enables / disables automatic freezing of the trees produces. By default enabled. | [Freezing](./freezing.mdx) |
 | `setUseProxies` | Can be used to disable or force the use of `Proxy` objects. Useful when filing bug reports. |  |
```
Comment (mweststrate) [inline website/docs/api.md:32]: TODO: remove

**Comment by markerikson** [comment]
Hey, just saw this!  How's it looking so far?

**Comment by mweststrate** [comment]
Mostly done except for docs / migration guide. I expect to release late
this week / next week

**Comment by markerikson** [comment]
Nice!

Any details on perf or size changes?

**Comment by markerikson** [comment]
Just ran a quick perf test with an RTK Query perf stress test scenario I've used previously.

I set it to render 1000 RTKQ-connected list items, then invalidated all the cache entries to force them to reload.

RTKQ dispatches when subscriptions are initiated, and again as data comes in.

With RTK 2.0.0-alpha.2 and Immer 9.x, average dispatch time during the load phase was 1.5ms.  With an RTK CSB CI build and Immer 10.0-beta.1, average dispatch time was 0.12ms (which is the lower limit of Chrome's function timing resolution):

image

When I hit "Invalidate", the Immer 9.x build had a solid 2.93 _seconds_ of dispatching.

With Immer 10, it was "only" 930ms:

image

The only change there was the Immer 10 bump.

So, clearly some pretty significant perf improvements!

That said, the Vite test project I'm using went _up_ in size with Immer 10.  The bundle went from 386K min (Immer 9.x) to 392K min (Immer 10).  Haven't looked into that yet, but very surprising.

**Comment by mweststrate** [comment]
Those are awesome results! Better than in my (more artificial?) benchmark. Size increase is correct, because Map/Set usage is no longer opt-in but enabled by default. Let me know if you think that'd be a problem, but we had quite some issues related to people getting confuses on how to enable it for RTK (iirc) and other packages, as they accidentally imported immer a second time for example just to enable map/set, or already used map/set on the critical path, before the enabling method was called. So figured this is more frictionless, and most projects use Maps sooner or later anyway in practice

**Comment by markerikson** [comment]
Just ran a quick check on 9.0.16 vs 10.0.beta, using Vite 4.2.1 + `source-map-explorer`:

- 9.0.16: `immer/dist/immer.esm.mjs`, 10.42KB
- 10.0.0-beta.4: `immer/dist/immer.mjs`, 15.78KB

If I hand-inspect the Vite bundle output, I don't _see_ any `NODE_ENV` references left in there, which is what I would expect - they exist in the ESM artifact, but the minifier should get rid of those.

If it helps at all, I just manually copied and pasted the minified versions of 9.0.16 and 10.0.0-beta.4 out of Vite-built bundles, ran them through Prettier, and am attaching them as a zip - maybe it'll help show where the differences are:


[immer-minified-prettified.zip](https://github.com/immerjs/immer/files/11140726/immer-minified-prettified.zip)



Per the `Map/Set` thing: FWIW, we really _don't_ want people using the `Map/Set` support with RTK because of serialization issues, so at least in our case that is a size increase we don't benefit from.  But, I do understand how that can cause confusion and annoyance on _your_ side.  (My own preference would be that it still be opt-in somehow if possible.)

**Comment by mweststrate** [comment]
Ah interesting! Is there a documentation section I can refer people back to
if they report this in the future? RTK is probably the biggest packaged
Immer provider, so if that isn't even supposed to be done in that case,
happy to reconsider

**Comment by markerikson** [comment]
Yeah, the two most immediate references are here:

- https://redux.js.org/style-guide/#do-not-put-non-serializable-values-in-state-or-actions
- https://redux.js.org/faq/organizing-state#can-i-put-functions-promises-or-other-non-serializable-items-in-my-store-state

We actually had an internal discussion about `Map/Set` usage again yesterday after seeing your comment, and concluded it's still not sufficiently worth it.  It _is_ hypothetically possible to rework the Redux DevTools with one of the new "JSON superset" serialization libs, but it's fairly common for folks to want to persist state in `localStorage` (often using the `redux-persist` library), and that would run into serialization issues.  There'd also be more nuance to try to explain in the docs, and it's simpler if we leave it at a blanket "Don't!" :)

**Comment by mweststrate** [comment]
@markerikson I reverted the MapSet opt-in in immer@10.0.0-beta.6, mind doing a check if it solves the bundle size increase?

**Comment by markerikson** [comment]
@mweststrate yeah, I'll check on that tonight!

**Comment by markerikson** [comment]
@mweststrate: Okay, 3 builds of the same RTKQ app, changing only the Immer version:

| Version     | Bundle (min) | Bundle (gzip) | Sourcemapped file |
|-------------|--------------|---------------|-------------------|
| 9.0.16      | 386.77 kB    | 122.41 kB     | 10.42 kB          |
| 10.0-beta.4 | 392.30 kB    | 123.74 kB     | 15.78 kB          |
| 10.0-beta.6 | 387.12 kB    | 122.31 kB     | 10.76 kB          |

**Comment by markerikson** [comment]
@mweststrate : well, you got me hooked :)

I cloned the `immer-10` branch and did some poking around. I think there's a few things going on:

- You're currently building the project with TSDX.  That seems to be adding some polyfills, like `_extends()` (which you can see in the CJS dev build).
- That in turn is partly because `tsconfig.json` is set to `target: "ES6"`.  

So, those are adding up to a bunch of dead bytes that aren't needed.

Unrelated to the bundle sizes, there's also issues with the current build output in this beta:

- I know you've been flipping file names back and forth. Right now there's an `immer.mjs` bundle, but an `immer.esm.js.map` sourcemap file
- The sourcemaps themselves seem broken and incorrect.  If you try loading some of them into https://evanw.github.io/source-map-visualization/ , you'll see that variables and segments don't line up correctly, making them semi-useless

FWIW, I _just_ switched `redux` and `redux-thunk` over to use https://tsup.egoist.dev/ .  It's basically like TSDX conceptually, except that it's based on ESBuild + Rollup (and it seems to actually be _maintained_ atm).

I just tried making a local branch that switches over to `tsup` instead, and after 15-ish minutes I've got a decent part of the output bundles generating.  More than that, I've got the minified files shrinking noticeably.  In `beta.6`, `immer.cjs.production.min.js` is 16542b, and with `tsup` I've already got it down to 13680b.  I also tried generating a minified `.mjs` artifact for comparison, and that's down to 12941b.

Two other notes here:

- I see room for some byte shaving tricks.  For example, I tried `export const getPrototypeOf = Object.getPrototypeOf` and deleting the `Object.` in front of the uses, and that shaved off a few bytes because the variable name gets minified down to just `E`.
- I'm 95% set on dropping UMD bundles for Redux 5.0 / RTK 2.0, especially since you can make a pre-built ESM bundle that should be importable in-browser.  Might want to consider that here.

Sooooo, with all that said:

If you're interested, **I can try to do the rest of the TSDX -> `tsup` conversion, and file that as a PR against this branch**.

Thoughts?

**Comment by mweststrate** [comment]
@markerikson thanks for pointing out tsup! I was already wondering what was the modern bundling setup since tsdx went unmaintained, which was also the reason I didn't want to poke too much in the config. So happy to receive a PR! Please do leave the flow types, as they're useful inside Meta, so I try to keep them roughly up to date :)

**Comment by markerikson** [comment]
@mweststrate okay, sure!  I'll try to put up a PR over the weekend.

**Comment by markerikson** [comment]
@mweststrate done! See #1032 .

Please let me know if you have any questions or would like me to tweak things further!

**Comment by mweststrate** [comment]
Sorry had some PTO and now playing really hard catch-up! I'll try to go
over it in the weekend :)

**Comment by mweststrate** [comment]
@markerikson just published `immer@10.0.0-beta.7`

**Pull request merged**